rust/lance-index/src/scalar/zoned.rs

```diff
@@ -0,0 +1,890 @@
+    /// the `_rowaddr` column with physical row addresses. Future zone-based
+    /// indexes should maintain this ordering or extend the trainer to accept an
+    /// explicit column index.
+    pub async fn train(
+        mut self,
+        stream: SendableRecordBatchStream,
+    ) -> Result<Vec<P::ZoneStatistics>> {
+        let zone_size = usize::try_from(self.zone_capacity).map_err(|_| {
+            Error::invalid_input(
+                "zone capacity does not fit into usize on this platform",
+                location!(),
+            )
+        })?;
+
+        let mut batches = chunk_concat_stream(stream, zone_size);
+        let mut zones = Vec::new();
+        let mut current_fragment_id: Option<u64> = None;
+        let mut current_zone_len: usize = 0;
+        let mut zone_start_offset: Option<u64> = None;
+        let mut zone_end_offset: Option<u64> = None;
+
+        self.processor.reset()?;
+
+        while let Some(batch) = batches.try_next().await? {
+            if batch.num_rows() == 0 {
+                continue;
+            }
+
+            let values = batch.column(0);
+            let row_addr_col = batch
+                .column_by_name(ROW_ADDR)
+                .ok_or_else(|| Error::invalid_input("missing _rowaddr column", location!()))?
+                .as_any()
+                .downcast_ref::<UInt64Array>()
+                .ok_or_else(|| {
+                    Error::invalid_input("_rowaddr column must be UInt64", location!())
+                })?;
+
+            let mut batch_offset = 0usize;
+            while batch_offset < batch.num_rows() {
+                let row_addr = row_addr_col.value(batch_offset);
+                let fragment_id = row_addr >> 32;
+
+                // Zones cannot span multiple fragments, flush when crossing boundary
+                match current_fragment_id {
+                    Some(current) if current != fragment_id => {
+                        Self::flush_zone(
+                            &mut self.processor,
+                            &mut zones,
+                            current,
+                            &mut current_zone_len,
+                            &mut zone_start_offset,
+                            &mut zone_end_offset,
+                        )?;
+                        current_fragment_id = Some(fragment_id);
+                    }
+                    None => {
+                        current_fragment_id = Some(fragment_id);
+                    }
+                    _ => {}
+                }
+
+                let run_len = Self::count_run_length(
+                    row_addr_col,
+                    batch_offset,
+                    batch.num_rows(),
+                    fragment_id,
+                );
+                let capacity = zone_size - current_zone_len;
+                let take = run_len.min(capacity);
+
+                if take == 0 {
+                    // Should not happen because flush occurs before reaching here,
```

Comment:
Is it trying to cover the [desired ==0](https://github.com/lance-format/lance/blob/main/rust/lance-index/src/scalar/bloomfilter.rs#L1079-L1087) logic? That's a valid condition. If not then delete the guard here as the possible infinite loop condition should be fixed in its root cause.